module/zcommon/zfs_fletcher_aarch64_neon.c

```diff
@@ -0,0 +1,231 @@
+register unsigned char ACC0 asm("v1") __attribute__((vector_size(16)));
+register unsigned char ACC1 asm("v2") __attribute__((vector_size(16)));
+register unsigned char ACC2 asm("v3") __attribute__((vector_size(16)));
+register unsigned char ACC3 asm("v4") __attribute__((vector_size(16)));
+register unsigned char TMP1 asm("v5") __attribute__((vector_size(16)));
+register unsigned char TMP2 asm("v6") __attribute__((vector_size(16)));
+register unsigned char SRC asm("v7") __attribute__((vector_size(16)));
+#else
+unsigned char ZERO __attribute__((vector_size(16)));
+unsigned char ACC0 __attribute__((vector_size(16)));
+unsigned char ACC1 __attribute__((vector_size(16)));
+unsigned char ACC2 __attribute__((vector_size(16)));
+unsigned char ACC3 __attribute__((vector_size(16)));
+unsigned char TMP1 __attribute__((vector_size(16)));
+unsigned char TMP2 __attribute__((vector_size(16)));
+unsigned char SRC __attribute__((vector_size(16)));
+#endif
+
+	asm("eor %[ZERO].16b,%[ZERO].16b,%[ZERO].16b\n"
+		"ld1 { %[ACC0].4s }, %[CTX0]\n"
+		"ld1 { %[ACC1].4s }, %[CTX1]\n"
+		"ld1 { %[ACC2].4s }, %[CTX2]\n"
+		"ld1 { %[ACC3].4s }, %[CTX3]\n"
+		: [ZERO] "=w" (ZERO),
+		[ACC0] "=w" (ACC0), [ACC1] "=w" (ACC1),
+		[ACC2] "=w" (ACC2), [ACC3] "=w" (ACC3)
+		: [CTX0] "Q" (ctx->aarch64_neon[0].v[0]),
+		[CTX1] "Q" (ctx->aarch64_neon[1].v[0]),
+		[CTX2] "Q" (ctx->aarch64_neon[2].v[0]),
+		[CTX3] "Q" (ctx->aarch64_neon[3].v[0]));
+
+	for (; ip < ipend; ip += 2) {
+		asm("ld1 { %[SRC].4s }, %[IP]\n"
+			"zip1 %[TMP1].4s, %[SRC].4s, %[ZERO].4s\n"
+			"zip2 %[TMP2].4s, %[SRC].4s, %[ZERO].4s\n"
+			"add %[ACC0].2d, %[ACC0].2d, %[TMP1].2d\n"
+			"add %[ACC1].2d, %[ACC1].2d, %[ACC0].2d\n"
+			"add %[ACC2].2d, %[ACC2].2d, %[ACC1].2d\n"
+			"add %[ACC3].2d, %[ACC3].2d, %[ACC2].2d\n"
+			"add %[ACC0].2d, %[ACC0].2d, %[TMP2].2d\n"
+			"add %[ACC1].2d, %[ACC1].2d, %[ACC0].2d\n"
+			"add %[ACC2].2d, %[ACC2].2d, %[ACC1].2d\n"
+			"add %[ACC3].2d, %[ACC3].2d, %[ACC2].2d\n"
+			: [SRC] "=&w" (SRC),
+			[TMP1] "=&w" (TMP1), [TMP2] "=&w" (TMP2),
+			[ACC0] "+w" (ACC0), [ACC1] "+w" (ACC1),
+			[ACC2] "+w" (ACC2), [ACC3] "+w" (ACC3)
+			: [ZERO] "w" (ZERO), [IP] "Q" (*ip));
+	}
+	asm("st1 { %[ACC0].4s },%[DST0]\n"
+		"st1 { %[ACC1].4s },%[DST1]\n"
+		"st1 { %[ACC2].4s },%[DST2]\n"
+		"st1 { %[ACC3].4s },%[DST3]\n"
+		: [DST0] "=Q" (ctx->aarch64_neon[0].v[0]),
+		[DST1] "=Q" (ctx->aarch64_neon[1].v[0]),
+		[DST2] "=Q" (ctx->aarch64_neon[2].v[0]),
+		[DST3] "=Q" (ctx->aarch64_neon[3].v[0])
+		: [ACC0] "w" (ACC0), [ACC1] "w" (ACC1),
+		[ACC2] "w" (ACC2), [ACC3] "w" (ACC3));
+}
+
+static void
+fletcher_4_aarch64_neon_byteswap(fletcher_4_ctx_t *ctx,
+    const void *buf, uint64_t size)
+{
+	const uint64_t *ip = buf;
+	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+	uint64_t v0[2], v1[2], v2[2], v3[2];
+	uint64_t A, B, C, D;
+#if defined(_KERNEL)
+register unsigned char ZERO asm("v0") __attribute__((vector_size(16)));
```

Comment:
This is a simple method that does not call into stuff. Could you just put used simd regs in the clobber list instead of these declarations? Would save a ton of space...